akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server.directives
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.{ HostConnectionPool, OutgoingConnection }
+import akka.http.scaladsl.model.Uri.Authority
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.model.{ HttpResponse, _ }
+import akka.http.scaladsl.server._
+import akka.http.scaladsl.server.directives.BasicDirectives._
+import akka.http.scaladsl.server.directives.FutureDirectives._
+import akka.http.scaladsl.server.directives.RouteDirectives._
+import akka.pattern.CircuitBreaker
+import akka.stream._
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+
+import scala.collection.immutable.Seq
+import scala.concurrent.Future
+import scala.util.{ Failure, Success, Try }
+
+trait ReverseProxyDirectives {
+  type RequestExecutor = ((HttpRequest) ⇒ Future[HttpResponse])
+
+  def completeUnmatchedPathWithReverseProxy(authority: Authority, configOption: Option[ReverseProxyConfig] = None): Route = {
+    forward(authority, configOption).tapply(x ⇒ complete(x._1))
+  }
+
+  def forward(authority: Authority, configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    extractUri.flatMap { uri ⇒
+      val forwardUri: Uri = uri.withAuthority(authority)
+      forwardToUri(forwardUri, configOption)
+    }
+  }
+
+  def forwardUnmatchedPath(authority: Authority, configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    extractUnmatchedPath.flatMap { unmatched ⇒
+      extractUri.flatMap { uri ⇒
+        val forwardUri: Uri = uri.withAuthority(authority).withPath(unmatched)
+        forwardToUri(forwardUri, configOption)
+      }
+    }
+  }
+
+  def forwardToUri(uri: Uri, configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    val requestUriInterceptor = ProxyUriRequestInterceptor(uri).andThen(ProxyHeaderRequestInterceptor)
+    mapRequest(requestUriInterceptor).tflatMap { _ ⇒
+      forwardRequest(configOption)
+    }
+  }
+
+  def forwardRequest(configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    extractExecutionContext.flatMap { implicit ec ⇒
+      extractRequest.flatMap { request ⇒
+        getOrCreateDefaultConfig(configOption).flatMap { reverseProxyConfig ⇒
+          val responseFuture: Future[HttpResponse] = reverseProxyConfig.requestExecutor.apply(request)
```

Comment:
Because of the way akka-http times out, the HttpResponse won't be consumed if it takes too long.  Assuming you just want to use the existing timeouts;

    def processTimeout(proxyResult: Future[HttpResponse])(implicit ec: ExecutionContext, mat: Materializer): (HttpRequest) => HttpResponse = { req =>
      proxyResult.onSuccess {
        case proxyResponse => proxyResponse.entity.dataBytes.runWith(Sink.ignore)
      }

      HttpResponse(StatusCodes.ServiceUnavailable, entity =
        HttpEntity.Strict(ContentTypes.`application/json`,
          ByteString("""{"error":"The server was not able to produce a timely response to your request."}""")))
     }

     withRequestTimeoutResponse(processTimeout(responseFuture) {
       ...
     }